sa/sysvars.go

```diff
@@ -0,0 +1,350 @@
+	}
+
+	if _, found := mariaDBNumericTypes[name]; found {
+		if checkFloatOrIntRE.FindString(value) == value {
+			return nil
+		}
+		if checkImproperFloatOrIntRE.FindString(value) == value {
+			return fmt.Errorf("%v=%v numeric is quoted, but should not be", name, value)
+		}
+	}
+
+	// Certain MariaDB enums can have both string and integer values.
+	mariaDBIntEnumTypes := map[string]struct{}{
+		"completion_type": {},
+	}
+
+	mariaDBStringEnumTypes := map[string]struct{}{
+		"completion_type":                {},
+		"default_storage_engine":         {},
+		"default_tmp_storage_engine":     {},
+		"optimizer_trace":                {},
+		"session_track_transaction_info": {},
+		"tx_isolation":                   {},
+	}
+
+	// Check the list of currently known MariaDB enumeration type system
+	// variables and determine if the value is either:
+	// 1) A properly formatted integer e.g. completion_type=1
+	if _, found := mariaDBIntEnumTypes[name]; found {
+		if checkIntRE.FindString(value) == value {
+			return nil
+		}
+		if checkImproperIntRE.FindString(value) == value {
+			return fmt.Errorf("%v=%v integer enum is quoted, but should not be", name, value)
+		}
+	}
+
+	// 2) A properly formatted string e.g. completion_type='CHAIN'
+	if _, found := mariaDBStringEnumTypes[name]; found {
+		if checkStringQuoteRE.FindString(value) != value {
+			return fmt.Errorf("%v=%v string enum is not properly quoted", name, value)
+		}
+		return nil
+	}
+
+	// MariaDB booleans can be (0, false) or (1, true).
+	// https://mariadb.com/kb/en/boolean/
+	mariaDBBooleanTypes := map[string]struct{}{
+		"autocommit":                   {},
+		"check_constraint_checks":      {},
+		"foreign_key_checks":           {},
+		"in_transaction":               {},
+		"keep_files_on_create":         {},
+		"log_slow_query":               {},
+		"low_priority_updates":         {},
+		"old":                          {},
+		"old_passwords":                {},
+		"profiling":                    {},
+		"query_cache_strip_comments":   {},
+		"query_cache_wlock_invalidate": {},
+		"session_track_schema":         {},
+		"session_track_state_change":   {},
+		"slow_query_log":               {},
+		"sql_auto_is_null":             {},
+		"sql_big_selects":              {},
+		"sql_buffer_result":            {},
+		"sql_if_exists":                {},
+		"sql_log_off":                  {},
+		"sql_notes":                    {},
+		"sql_quote_show_create":        {},
+		"sql_safe_updates":             {},
+		"sql_warnings":                 {},
+		"standard_compliant_cte":       {},
+		"tcp_nodelay":                  {},
+		"tx_read_only":                 {},
+		"unique_checks":                {},
+		"updatable_views_with_limit":   {},
+	}
+
+	if _, found := mariaDBBooleanTypes[name]; found {
+		if checkBooleanRE.FindString(value) != value {
+			return fmt.Errorf("%v=%v expected boolean value", name, value)
+		}
+		return nil
+	}
+
+	return fmt.Errorf("%v=%v was unexpected", name, value)
```

Comment:
I was reading through this PR, thinking to myself "hmm, how do we make sure that the contents of all of the individual string/int/etc type maps add up to exactly the contents of the allowed session varaibles map?". And this line here provides my answer: we _shouldn't_.

I suggest we remove the mariaDBSessionScope map entirely, and rely on this line to error out any key-value pairs which aren't recognized by one of the other maps. And then very clearly document that each map should only contain session-scoped vars, DO NOT add global-scope int vars to this map, etc etc. Does that seem reasonable?